src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -0,0 +1,338 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "command_line_parser.h"
+#include "common.h"
+#include "doctest.h"
+#include "mock_client_backend.h"
+#include "mock_model_parser.h"
+#include "request_rate_manager.h"
+
+namespace cb = triton::perfanalyzer::clientbackend;
+
+namespace triton { namespace perfanalyzer {
+
+class TestRequestRateManager {
+  public:
+    TestRequestRateManager() {
+      // Must reset this global flag every unit test. 
+      // It gets set to true when we deconstruct any load manager 
+      // (aka every unit test)
+      //
+      early_exit = false;
+
+      stats_ = std::make_shared<cb::MockClientStats>();
+    }
+
+    /// Test that the correct Infer function is called in the backend
+    ///
+    void TestInferType(bool is_async, bool is_streaming) {
+      PerfAnalyzerParameters params;
+      params.async = is_async;
+      params.streaming = is_streaming;
+
+      double request_rate = 50;
+      auto sleep_time = std::chrono::milliseconds(100);
+
+      std::unique_ptr<LoadManager> manager = CreateManager(params);
+      dynamic_cast<RequestRateManager*>(manager.get())->ChangeRequestRate(request_rate);
+      std::this_thread::sleep_for(sleep_time);
+
+      CheckInferType(params);
```

Comment:
Yes, it probably could have been done that way. It would have involved modifications to the class to provide access to the function. After we refactor hopefully there will be a much cleaner place to test this. For now I prefer end-to-end testing. I fully expect this functionality to be pulled out into its own class.

Reply:
That's fine if we wanna leave it as is and fix it up later. For what it's worth, though, I think the modifications to the class would be quite limited. Probably just making the particular testing class a friend/subclass of the RequestRateManager so that it has access to private parent-class methods to call. Mocking the client backend stuff was already done somewhere else in the PR so that's not really a new modification to RequestRateManager specific to this thread.

Reply:
I agree this one wouldn't have been that bad to do. I don't expect these tests to stay for long. It should end up being a whole new class and can be done properly there.